src/io/delimited.cr

```diff
@@ -0,0 +1,124 @@
+module IO
+  # An IO that wraps another IO, and only reads up to the beginning of a
+  # specified delimiter.
+  #
+  # This is useful for exposing part of an underlying stream to a client.
+  #
+  # ```
+  # io = MemoryIO.new "abc||123"
+  # delimited = IO::Delimited.new(io, read_delimiter: "||")
+  #
+  # delimited.gets_to_end # => "abc"
+  # delimited.gets_to_end # => nil
+  # io.gets_to_end        # => "123"
+  # ```
+  class Delimited
+    include IO
+
+    # If `sync_close` is true, closing this IO will close the underlying IO.
+    property? sync_close
+
+    getter read_delimiter
+    getter? closed : Bool
+
+    @delimiter_buffer : Bytes
+    @active_delimiter_buffer : Bytes?
```

Comment:
@asterite but even in the case that `byte` NEVER equals `first_byte` (first run), it's nearly twice as fast. This means `active_delimiter_buffer` is always nil, and the only changed code that actually runs is lines 58-63. Even then, `if buffer = @active_delimiter_buffer` is always false. This doesn't seem to support your first hypothesis, that it's more complex therefore slower, because to me thats almost the same number of operations, which means either the operations are more expensive, or that the codegen is radically different in that path.

I'm not too well versed in LLVM internals, but your second theory seems much more likely.